File: backend/src/main/java/com/matriculaonline/service/DisciplinaService.java
package com.matriculaonline.service;

import com.matriculaonline.domain.exception.ResourceNotFoundException;
import com.matriculaonline.domain.model.Curso;
import com.matriculaonline.domain.model.Disciplina;
import com.matriculaonline.dto.request.DisciplinaRequest;
import com.matriculaonline.dto.response.DisciplinaResponse;
import com.matriculaonline.dto.response.PageResponse;
import com.matriculaonline.repository.CursoRepository;
import com.matriculaonline.repository.DisciplinaRepository;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

@Service
public class DisciplinaService {

    private final DisciplinaRepository disciplinaRepository;
    private final CursoRepository cursoRepository;

    public DisciplinaService(DisciplinaRepository disciplinaRepository, CursoRepository cursoRepository) {
        this.disciplinaRepository = disciplinaRepository;
        this.cursoRepository = cursoRepository;
    }

    @Transactional
    public DisciplinaResponse criar(DisciplinaRequest request) {
        Curso curso = cursoRepository.findByUuid(request.cursoUuid())
                .orElseThrow(() -> new ResourceNotFoundException("Curso", request.cursoUuid().toString()));

        Disciplina disciplina = new Disciplina();
        disciplina.setNome(request.nome());
        disciplina.setDescricao(request.descricao());
        disciplina.setCargaHoraria(request.cargaHoraria());
        disciplina.setCurso(curso);
        return DisciplinaResponse.fromEntity(disciplinaRepository.save(disciplina));
    }

    @Transactional(readOnly = true)
    public PageResponse<DisciplinaResponse> listar(Pageable pageable) {
        return PageResponse.from(disciplinaRepository.findAll(pageable), DisciplinaResponse::fromEntity);
    }

    @Transactional(readOnly = true)
    public DisciplinaResponse buscarPorUuid(UUID uuid) {
        Disciplina disciplina = disciplinaRepository.findByUuid(uuid)
                .orElseThrow(() -> new ResourceNotFoundException("Disciplina", uuid.toString()));
        return DisciplinaResponse.fromEntity(disciplina);
    }

    @Transactional
    public DisciplinaResponse atualizar(UUID uuid, DisciplinaRequest request) {
        Disciplina disciplina = disciplinaRepository.findByUuid(uuid)
                .orElseThrow(() -> new ResourceNotFoundException("Disciplina", uuid.toString()));

        Curso curso = cursoRepository.findByUuid(request.cursoUuid())
                .orElseThrow(() -> new ResourceNotFoundException("Curso", request.cursoUuid().toString()));

        disciplina.setNome(request.nome());
        disciplina.setDescricao(request.descricao());
        disciplina.setCargaHoraria(request.cargaHoraria());
        disciplina.setCurso(curso);
        return DisciplinaResponse.fromEntity(disciplinaRepository.save(disciplina));
    }

    /**
     * Remove a disciplina. Idempotente: se o recurso já não existir, retorna sem erro (no-op).
     */
    @Transactional
    public void deletar(UUID uuid) {
        disciplinaRepository.findByUuid(uuid).ifPresent(disciplinaRepository::delete);
    }
}

File: frontend/src/app/services/disciplina.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Disciplina, DisciplinaRequest } from '../models/disciplina.model';
import { PageResponse } from '../models/page.model';

@Injectable({ providedIn: 'root' })
export class DisciplinaService {
  private readonly apiUrl = '/api/disciplinas';

  constructor(private http: HttpClient) {}

  listar(page = 0, size = 10, sort = 'nome,asc'): Observable<PageResponse<Disciplina>> {
    const params = new HttpParams()
      .set('page', page)
      .set('size', size)
      .set('sort', sort);
    return this.http.get<PageResponse<Disciplina>>(this.apiUrl, { params });
  }

  buscarPorUuid(uuid: string): Observable<Disciplina> {
    return this.http.get<Disciplina>(`${this.apiUrl}/${uuid}`);
  }

  criar(request: DisciplinaRequest): Observable<Disciplina> {
    return this.http.post<Disciplina>(this.apiUrl, request);
  }

  atualizar(uuid: string, request: DisciplinaRequest): Observable<Disciplina> {
    return this.http.put<Disciplina>(`${this.apiUrl}/${uuid}`, request);
  }

  deletar(uuid: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${uuid}`);
  }
}
